[김은혜] Step4 구간 추가 제거 PR 요청

eunveloper opened the pull request:
안녕하세요 리뷰어님!
길고도 어려웠던(?) 이번 주차에 마지막 단계네요 ㅠㅠ

저번 PR에서 제가 초기에 방향을 너무 잘못잡고 그러다 보니 상당히 까다로우셨을텐데
끝까지 꼼꼼히 봐주셔서 감사합니다 🥲

그래서 덕분에 정말정말 많이 배웠어요!
마지막까지 잘 부탁드립니다!

eunveloper committed:
refactor: 3단계 피드백 반영

- 기능 이관 (DTO -> Domain)
- 기능 버그 수정
  - 상행종점 찾는기능
  - 상행 기준으로 하행 찾는 기능

eunveloper committed:
docs: README.md 작성

eunveloper committed:
feat: 테스트코드 작성

- 종점을 제거
- 가운데 역을 제거
- 단일구간의 노선에 역을 제거
- 노선에 포함되지 않는 역을 제거

eunveloper committed:
feat: 비지니스 로직 개발

- 노선 내 역 제거 로직 추가
- 각종 예외처리 추가

geunhwanlee commented on src/main/java/nextstep/subway/domain/Section.java:
```diff
@@ -93,14 +89,39 @@ public boolean isDownStationEqualsId(Station station) {
-    public boolean containsAllStation(Section infixSection) {
-        if ((upStation.equalsById(infixSection.upStation) && downStation.equalsById(infixSection.downStation))
-                || (upStation.equalsById(infixSection.downStation) && downStation.equalsById(infixSection.upStation))) {
+    public boolean isContainsPreStation(boolean isContains, Section infixSection) {
+        if (isContains) {
+            return true;
+        }
```
`isContains`가 하는 역할이 이 메서드만 보면 잘 드러나지 않네요. 😭
이 메서드에서는 `isContains` 없이 처리하고 이 메서드 외부에서 로직을 구현하는건 어떨까요? 🤔

eunveloper replied:
메소드명의 행위만 하도록 `isContainsPreStation` 직전역에 속해있는지 여부를 판단하는 로직만 남겨놓고
이전에 이미 속해있는 필드값을 비교하여 containStatus 를 변경하는 메소드를 다시 구성하여 분리하였습니다!

geunhwanlee commented on src/main/java/nextstep/subway/domain/Sections.java:
```diff
@@ -1,18 +1,25 @@
 public class Sections {
 
-    @OneToMany(fetch = FetchType.LAZY, mappedBy = "line", cascade = CascadeType.ALL)
+    @OneToMany(
+            fetch = FetchType.LAZY,
+            mappedBy = "line",
+            cascade = CascadeType.ALL,
+            orphanRemoval = true)
```
`orphanRemoval` 활용 👍🏻

geunhwanlee commented on src/main/java/nextstep/subway/domain/Sections.java:
```diff
@@ -26,12 +33,68 @@ public void infix(Section infixSection) {
+    }
+
+    private boolean isHaveOnlyOneSection() {
+        if (sections.size() < 2) {
+            return true;
+        }
```
2라는 값이 무엇을 의미하는지 표현되지 않았네요.
이름을 부여하면 어떨까요? 😀

[의미가 불분명한 매직 넘버를 상수로 선언하라](https://javabom.tistory.com/28)

eunveloper replied:
오 안그래도 이번 기능 구현하면서
뭔가 의미가 내포된 로직들을 어떻게하면 좀더 의미가 보일수 있을까 고민해서
메소드명이나 변수명으로 풀려고 좀 노력을 했었는데,
이런방식이 있었군요!!!!! 👍🏻👍🏻👍🏻👍🏻👍🏻

geunhwanlee commented on src/main/java/nextstep/subway/domain/Sections.java:
```diff
@@ -26,12 +33,68 @@ public void infix(Section infixSection) {
+    private void deleteStation(List<Section> sectionsContainStation, Station station) {
+        if (sectionsContainStation.size() == 0) {
+            throw new InvalidParameterException("호선에 존재하지 않는 역은 제거할 수 없습니다.");
+        }
```
```suggestion
        if (sectionsContainStation.isEmpty()) {
            throw new InvalidParameterException("호선에 존재하지 않는 역은 제거할 수 없습니다.");
        }
```
이렇게 써볼수 있겠네요 😀

geunhwanlee commented on src/main/java/nextstep/subway/domain/Sections.java:
```diff
@@ -26,12 +33,68 @@ public void infix(Section infixSection) {
+            throw new InvalidParameterException("호선에 존재하지 않는 역은 제거할 수 없습니다.");
+        }
+
+        if (sectionsContainStation.size() == 2) {
+            sections.add(createDeleteSection(sectionsContainStation, station));
+            sections.remove(sectionsContainStation.get(1));
+        }
+        sections.remove(sectionsContainStation.get(0));
```
`sectionsContainStation.size()`가 1이라면 상행종점역 또는 하행종점역을 지우는경우고
2라면 중간역을 지우는경우일거같아요! 😀
이런 의미가 코드에 담기도록 개선해볼수 있지 않을까요?

eunveloper replied:
위 댓글처럼
뭔가 내재된 의미를 확실하게 보여주는 방식이 다양할거 같네요!
좀더 고민해서 수정해보겠습니다 ㅎㅎ

geunhwanlee commented on src/main/java/nextstep/subway/domain/Sections.java:
```diff
@@ -26,12 +33,68 @@ public void infix(Section infixSection) {
+        return sectionsContainStation.get(0).delete(sectionsContainStation.get(1), station);
+    }
+
+    private List<Section> getSectionsContainStation(Station station) {
+        List<Section> sectionsContainStation = new ArrayList<>();
+        for (Section section : sections) {
+            sectionsContainStation.add(getSectionContainStation(section, station));
+            sectionsContainStation.remove(null);
+        }
+        return sectionsContainStation;
+    }
+
+    private Section getSectionContainStation(Section section, Station station) {
+        if (section.getUpStation().equalsById(station)) {
+            return section;
+        }
+        if (section.getDownStation().equalsById(station)) {
+            return section;
+        }
+        return null;
+    }
```
```suggestion
    private List<Section> getSectionsContainStation(Station station) {
        return sections.stream()
                .filter(section -> section.contains(station))
                .collect(Collectors.toList());
    }
```
null을 `add`했다가 `remove`하기보다는
이렇게 작성해보는건 어떨까요?
`section.getUpstation()` 과 같은 getter도 제거할수 있고 로직도 파악하기 쉬운것 같은데 어떻게 생각하시나요?

eunveloper replied:
와ㅏ아아아아 이렇게도 구현해도 되겠네요 ㅠㅠ
훨씬 코드도 깔끔하고 section 내의 필드에 접근하는 행위를 section 도메인에서 구현할수 있어서
굉장히 좋은거 같아요!!!!!

eunveloper replied:
와 다시보니 정말 stream이 컴팩트하고 좋네요
sections 를 순회하면서 무언가의 조건에 충족된 데이터를 모아 리스트로 구현하는 행위하는 로직을

1. sections 를 순회하면서 `stream`
2. 무언가의 조건에 충족된 데이터를 모아 `filter`
3. 리스트로 구현 `collect`

하게 깔끔하게 구성이 되는.......👍🏻

geunhwanlee commented on src/main/java/nextstep/subway/domain/Sections.java:
```diff
@@ -91,4 +154,55 @@ private boolean isNothingToChange(Section changedStationSection) {
         return changedStationSection == null;
     }
 
+    public Set<StationResponse> sortSections() {
+        Set<StationResponse> stationResponses = new LinkedHashSet<>();
+
+        Section firstSection = findFirstSection();
+        stationResponses.add(StationResponse.of(firstSection.getUpStation()));
+        for (int i = 0; i < sections.size(); i++) {
+            firstSection = addedNextSectionStations(firstSection, stationResponses);
+        }
+
+        return stationResponses;
+    }
```
`StationResponse`에 대한 의존성이 생겼네요. 😭
`Station`으로 리턴해줘도 될거같아요.

geunhwanlee commented on src/main/java/nextstep/subway/domain/Sections.java:
```diff
@@ -91,4 +154,55 @@ private boolean isNothingToChange(Section changedStationSection) {
+        return firstSection;
+    }
+
+    private Section findFirstSection() {
+        Section firstSection = sections.get(0);
+        for (int i = 0; i < sections.size(); i++) {
+            firstSection = findPreSection(firstSection);
+        }
+        return firstSection;
+    }
+
+    private Section findPreSection(Section firstSection) {
+        for (Section section : sections) {
+            firstSection = resetFirstSection(section, firstSection);
+        }
+        return firstSection;
+    }
+
+    private Section resetFirstSection(Section section, Section firstSection) {
+        if (section.isPreStation(firstSection)) {
+            return section;
+        }
+        return firstSection;
+    }
```
`상행종점역을 찾는 로직`과 `하행역 방향으로 탐색하는 로직`을 파악하는데 힘들었네요. 😭

제가 파악한 바로는 `findPreSection`은 한번 실행될 때마다 `최소1칸에서 최대 sections의 크기만큼 상행역으로 이동`하는 메서드고
`findPreSection`을 `sections.size()`만큼 반복하는 이유는 최악의 케이스를 고려해서 최대한의 반복을 하기 위함이네요! 💡

이렇게 작성해도 충분히 잘 동작하겠네요! 👍🏻
다만 가독성이 조금 떨어지는것같아요.. 😭

**꼭 고치실 필요는 없습니다!**
다만 어떻게 하면 읽기 좋은 코드로 개선할수 있을지 고민해보시면 좋을것같아요! 🙏

eunveloper replied:
저 개인적으로도 이번 과제에서 가장 골머리 썩힌 부분인거 같아요 ㅠㅠ
제가 알고리즘 적으로도 부족하다보니 알고리즘을 풀어내는것도 급급한데 이걸 좀더 읽기 좋게 구현한다는게 꽤나 어렵더라구요 ㅠㅠ
else문을 사용하지 않고, 2뎁스 사용을 안하고 또 알고리즘 코드를 풀어가다보니
break 제어도 굉장히 까다롭고... 또 stream의 foreach() 는 break 제어가 안되고... stream 블럭 밖에 선언된 지역변수에 변경이 어렵고... 감안하고 atomic 객체를 써야하는게 맞나 싶기도 하고요 ㅠㅠ

리뷰어 님께도 고생 해주시고 꼼꼼히 봐주셔서 감사했습니다!!
좀더 공부하고 고민해서 개선하게 된다면 몰래 PR 날려보겠습니다.....ㅎ_ㅎ

geunhwanlee reviewed (changes requested):
4단계 미션 잘해주셨어요! 👍🏻

충분히 잘 해주셨지만
몇가지 고민해볼만한 내용이 있어서 코멘트 드렸어요.

한번 고민해보시고 재 리뷰요청 주세요! 🙏
그럼 화이팅하세요! 🔥

eunveloper committed:
refactor: 피드백 반영

- 메소드 분리 (메소드명의 행위만 하도록)
- 매직넘버 상수로 분리
- 내재된 의미 좀더 의미를 드러내도록 코드 수정
- 로직 리패토링
- 의존성 제거

geunhwanlee commented on src/main/java/nextstep/subway/domain/Sections.java:
```diff
@@ -26,12 +41,73 @@ public void infix(Section infixSection) {
+    }
+
+    private boolean isHaveOnlyOneSection() {
+        if (sections.size() < MANY_SECTION_COUNT) {
+            return true;
+        }
```
```suggestion
        if (sections.size() <= MIN_SIZE) {
            return true;
        }
```
이렇게 해보면 어떨까요? 😀

geunhwanlee commented on src/main/java/nextstep/subway/domain/Sections.java:
```diff
@@ -26,12 +41,73 @@ public void infix(Section infixSection) {
+        return false;
+    }
+
+    private void deleteStation(Sections sectionsContainStation, Station station) {
+        if (sectionsContainStation.isNotFoundStation()) {
+            throw new InvalidParameterException("호선에 존재하지 않는 역은 제거할 수 없습니다.");
+        }
+
+        if (sectionsContainStation.isDeleteInnerStation()) {
+            sections.add(createDeleteSection(sectionsContainStation, station));
+            sections.remove(sectionsContainStation.sections.get(1));
+        }
+        sections.remove(sectionsContainStation.sections.get(0));
+    }
+
+    private boolean isDeleteInnerStation() {
+        return sections.size() == 2;
+    }
+
+    private boolean isNotFoundStation() {
+        return sections.isEmpty();
+    }
```
의미를 담기위해 클래스로 묶고 메서드로 나타낸것 좋습니다! 👍🏻

다만 제 생각으론 `sectionsContainStation`는 비즈니스적으로 의미가 있는 값이 아닌 구현을 위해 임의로 만든 값이기 때문에 객체로 포장할 필요는 없는거같아요!
메서드로 추출하는정도만 해도 충분하지 않았나 싶어요. 😀

그래도 잘 해주셨습니다! 👏

geunhwanlee reviewed (approved):
리뷰 반영 잘 해주셨습니다! 👍🏻

충분히 잘 해주셔서 머지하도록 할게요. 😀

미션이 쉽지 않으셨을텐데 고생 많으셨고 잘 해주셨습니다. 👏

소소한 코멘트 남겼으니 확인해주세요!

그럼 앞으로의 미션도 화이팅하세요! 🔥
